File: xml_parser.py
"""
xml_parser.py
Parse file XML phiếu xuất kho kiêm vận chuyển nội bộ (hóa đơn điện tử Việt Nam).
"""

import xml.etree.ElementTree as ET


def _text(element, tag, default=""):
    """Lấy text của thẻ con, trả về default nếu không tìm thấy."""
    child = element.find(tag)
    if child is not None and child.text:
        return child.text.strip()
    return default


def parse_invoice(xml_file: str) -> dict:
    """
    Parse file XML hóa đơn điện tử Việt Nam.

    Args:
        xml_file: Đường dẫn đến file XML.

    Returns:
        dict với 2 key:
            - 'header': dict thông tin header hóa đơn
            - 'items': list các dict thông tin từng dòng hàng hóa
    """
    tree = ET.parse(xml_file)
    root = tree.getroot()

    # Loại bỏ namespace nếu có (ví dụ: {http://...}HDon → HDon)
    _strip_ns(root)

    # Tìm node chứa dữ liệu hóa đơn — hỗ trợ cả cấu trúc có DLHDon hoặc không
    _dl = root.find(".//DLHDon")
    dl_hdon = _dl if _dl is not None else root

    _ttc = dl_hdon.find(".//TTChung")
    if _ttc is None:
        _ttc = dl_hdon.find("TTChung")
    tt_chung = _ttc if _ttc is not None else ET.Element("TTChung")

    _nd = dl_hdon.find(".//NDHDon")
    if _nd is None:
        _nd = dl_hdon.find("NDHDon")
    nd_hdon = _nd if _nd is not None else dl_hdon

    _nb = nd_hdon.find("NBan")
    n_ban = _nb if _nb is not None else ET.Element("NBan")

    _nm = nd_hdon.find("NMua")
    n_mua = _nm if _nm is not None else ET.Element("NMua")

    header = {
        # Thông tin chung
        "SoHoaDon": _text(tt_chung, "SHDon") or _text(tt_chung, "SoHDon"),
        "NgayXuatHoaDon": _text(tt_chung, "NLap"),
        "MauSo": _text(tt_chung, "KHMSHDon"),
        "KyHieu": _text(tt_chung, "KHHDon"),
        "DongTienThanhToan": _text(tt_chung, "LoaiTien"),
        "MaLoaiHoaDon": _text(tt_chung, "MaLoaiHD") or _text(tt_chung, "MaLoai"),
        # Tên hóa đơn lấy từ TenHDon hoặc THHD (tên hóa đơn)
        "GhiChu": _text(tt_chung, "TenHDon") or _text(tt_chung, "THHD"),
        # Người bán
        "BenBanTenDonVi": _text(n_ban, "Ten"),
        "BenBanMaSoThue": _text(n_ban, "MST"),
        "BenBanDiaChi": _text(n_ban, "DChi"),
        "BenBanDienThoai": _text(n_ban, "SDThoai") or _text(n_ban, "SoDienThoai"),
        "BenBanFax": _text(n_ban, "Fax"),
        "BenBanTaiKhoanNganHang": _text(n_ban, "STKNHang") or _text(n_ban, "SoTK"),
        "BenBanTenNganHang": _text(n_ban, "TNHang") or _text(n_ban, "TenNganHang"),
        # Người mua
        "BenMuaTenDonVi": _text(n_mua, "Ten"),
        "BenMuaMaSoThue": _text(n_mua, "MST"),
        "BenMuaDiaChi": _text(n_mua, "DChi"),
        "BenMuaHoTen": _text(n_mua, "HoTen") or _text(n_mua, "NMHHoa"),
        "BenMuaDienThoai": _text(n_mua, "SDThoai") or _text(n_mua, "SoDienThoai"),
        "BenMuaEmail": _text(n_mua, "DCTDTu") or _text(n_mua, "Email"),
        "BenMuaTaiKhoanNganHang": _text(n_mua, "STKNHang") or _text(n_mua, "SoTK"),
        "BenMuaTenNganHang": _text(n_mua, "TNHang") or _text(n_mua, "TenNganHang"),
        "BenMuaMaDonVi": _text(n_mua, "MDVQHKhach") or _text(n_mua, "MaDonVi"),
        # Tổng tiền
        "TongTienHang": _parse_number(_text(nd_hdon, "TToan") or _text(nd_hdon, "TgTCThue")),
        "TienThueVat": _parse_number(_text(nd_hdon, "TgTThue")),
        "TongTienThanhToan": _parse_number(
            _text(nd_hdon, "TgTTToan") or _text(nd_hdon, "TTToan")
        ),
        "TongTienThanhToanBangChu": _text(nd_hdon, "TTTBChu") or _text(nd_hdon, "TienBangChu"),
        # Thông tin vận chuyển (phiếu xuất kho)
        "SoHopDongVanChuyen": _text(nd_hdon, "SoHDVChuyen") or _text(nd_hdon, "SoHopDong"),
        "HoTenNguoiVanChuyen": _text(nd_hdon, "HoTenNVChuyen"),
        "MaPhuongTienVanChuyen": _text(nd_hdon, "MPTVChuyen") or _text(nd_hdon, "BienSoXe"),
        "XuatTaiKho": _text(nd_hdon, "XuatTaiKho") or _text(nd_hdon, "DiemXuat"),
        "NhapTaiKho": _text(nd_hdon, "NhapTaiKho") or _text(nd_hdon, "DiemNhap"),
        "LenhDieuDongSo": _text(nd_hdon, "SoLDD") or _text(nd_hdon, "LenhDieuDong"),
        "HinhThucThanhToan": _text(nd_hdon, "HTTToan") or _text(nd_hdon, "HinhThucTT"),
        "isPXK": 1,  # Đây là phiếu xuất kho
    }

    # Parse danh sách hàng hóa dịch vụ
    items = []
    _ds = nd_hdon.find("DSHHDVu")
    ds_hhdvu = _ds if _ds is not None else nd_hdon
    for hhdvu in ds_hhdvu.findall("HHDVu"):
        item = {
            "TinhChat": _parse_int(_text(hhdvu, "TChat")),
            "SoThuTu": _parse_int(_text(hhdvu, "STT")),
            "MaHang": _text(hhdvu, "MHHDVu") or _text(hhdvu, "MaHang"),
            "TenHang": _text(hhdvu, "THHDVu") or _text(hhdvu, "TenHang"),
            "DonViTinh": _text(hhdvu, "DVTinh"),
            "SoLuong": _parse_number(_text(hhdvu, "SLuong")),
            "DonGia": _parse_number(_text(hhdvu, "DGia")),
            "ThanhTien": _parse_number(_text(hhdvu, "ThTien") or _text(hhdvu, "TgTKhau")),
            "ChietKhauTien": _parse_number(_text(hhdvu, "STCKhau")),
            "VAT": _parse_number(_text(hhdvu, "TSuat")),
            "TienVat": _parse_number(_text(hhdvu, "TienThue") or _text(hhdvu, "TTSuat")),
            "GhiChu": _text(hhdvu, "GhiChu"),
        }
        items.append(item)

    return {"header": header, "items": items}


def _strip_ns(element):
    """Đệ quy loại bỏ namespace khỏi tất cả tag trong cây XML."""
    if element.tag and element.tag.startswith("{"):
        element.tag = element.tag.split("}", 1)[1]
    for child in element:
        _strip_ns(child)


def _parse_number(value: str):
    """Chuyển chuỗi thành số thực; trả về None nếu rỗng hoặc không hợp lệ."""
    if not value:
        return None
    try:
        return float(value.replace(",", "."))
    except ValueError:
        return None


def _parse_int(value: str):
    """Chuyển chuỗi thành số nguyên; trả về None nếu rỗng hoặc không hợp lệ."""
    if not value:
        return None
    try:
        return int(value)
    except ValueError:
        return None

File: web/XmlParser.php
<?php
/**
 * XmlParser.php
 * Parse file XML hóa đơn điện tử Việt Nam — port từ xml_parser.py.
 */

class InvoiceXmlParser
{
    /**
     * Parse file XML hóa đơn điện tử.
     *
     * @param  string $xmlFilePath  Đường dẫn đến file XML.
     * @return array  ['header' => [...], 'items' => [...]]
     * @throws RuntimeException nếu không đọc / parse được file.
     */
    public static function parse(string $xmlFilePath): array
    {
        if (!file_exists($xmlFilePath)) {
            throw new RuntimeException("File XML không tồn tại: $xmlFilePath");
        }

        $content = file_get_contents($xmlFilePath);
        if ($content === false) {
            throw new RuntimeException("Không thể đọc file XML: $xmlFilePath");
        }

        // Tắt lỗi libxml để xử lý thủ công
        libxml_use_internal_errors(true);
        $xml = simplexml_load_string($content);
        if ($xml === false) {
            $errors = libxml_get_errors();
            libxml_clear_errors();
            $msg = implode('; ', array_map(fn($e) => trim($e->message), $errors));
            throw new RuntimeException("Lỗi parse XML: $msg");
        }
        libxml_clear_errors();

        // Xoá namespace — convert sang DOM rồi trở về SimpleXML
        $dom = dom_import_simplexml($xml)->ownerDocument;
        self::stripNamespaces($dom->documentElement);
        $root = simplexml_import_dom($dom->documentElement);

        // Tìm node chứa dữ liệu hóa đơn (hỗ trợ cả cấu trúc có/không có DLHDon)
        $dlHdon = $root->xpath('.//DLHDon');
        $dlHdon = !empty($dlHdon) ? $dlHdon[0] : $root;

        $ttChung = $dlHdon->xpath('.//TTChung');
        $ttChung = !empty($ttChung) ? $ttChung[0] : new SimpleXMLElement('<TTChung/>');

        $ndHdon = $dlHdon->xpath('.//NDHDon');
        $ndHdon = !empty($ndHdon) ? $ndHdon[0] : $dlHdon;

        $nBan = $ndHdon->NBan ?? new SimpleXMLElement('<NBan/>');
        $nMua = $ndHdon->NMua ?? new SimpleXMLElement('<NMua/>');

        // ---- Header ----
        $soHoaDon = self::text($ttChung, 'SHDon') ?: self::text($ttChung, 'SoHDon');
        $header = [
            // Thông tin chung
            'SoHoaDon'                  => $soHoaDon,
            'NgayXuatHoaDon'            => self::text($ttChung, 'NLap'),
            'MauSo'                     => self::text($ttChung, 'KHMSHDon'),
            'KyHieu'                    => self::text($ttChung, 'KHHDon'),
            'DongTienThanhToan'         => self::text($ttChung, 'LoaiTien'),
            'MaLoaiHoaDon'              => self::text($ttChung, 'MaLoaiHD') ?: self::text($ttChung, 'MaLoai'),
            'GhiChu'                    => self::text($ttChung, 'TenHDon') ?: self::text($ttChung, 'THHD'),
            // Người bán
            'BenBanTenDonVi'            => self::text($nBan, 'Ten'),
            'BenBanMaSoThue'            => self::text($nBan, 'MST'),
            'BenBanDiaChi'              => self::text($nBan, 'DChi'),
            'BenBanDienThoai'           => self::text($nBan, 'SDThoai') ?: self::text($nBan, 'SoDienThoai'),
            // Người mua
            'BenMuaTenDonVi'            => self::text($nMua, 'Ten'),
            'BenMuaMaSoThue'            => self::text($nMua, 'MST'),
            'BenMuaDiaChi'              => self::text($nMua, 'DChi'),
            // Tổng tiền
            'TongTienHang'              => self::parseNumber(
                self::text($ndHdon, 'TToan') ?: self::text($ndHdon, 'TgTCThue')
            ),
            'TienThueVat'               => self::parseNumber(self::text($ndHdon, 'TgTThue')),
            'TongTienThanhToan'         => self::parseNumber(
                self::text($ndHdon, 'TgTTToan') ?: self::text($ndHdon, 'TTToan')
            ),
            // Thông tin vận chuyển
            'SoHopDong'                 => self::text($ndHdon, 'SoHDVChuyen') ?: self::text($ndHdon, 'SoHopDong'),
            'HinhThucThanhToan'         => self::text($ndHdon, 'HTTToan') ?: self::text($ndHdon, 'HinhThucTT'),
        ];

        // ---- Items ----
        $items = [];
        $dsHHDVu = $ndHdon->DSHHDVu ?? $ndHdon;
        $hhdvuList = $dsHHDVu->HHDVu ?? [];
        foreach ($hhdvuList as $hh) {
            $items[] = [
                'TinhChat'   => self::parseInt(self::text($hh, 'TChat')),
                'SoThuTu'   => self::parseInt(self::text($hh, 'STT')),
                'MaHang'    => self::text($hh, 'MHHDVu') ?: self::text($hh, 'MaHang'),
                'TenHang'   => self::text($hh, 'THHDVu') ?: self::text($hh, 'TenHang'),
                'DonViTinh' => self::text($hh, 'DVTinh'),
                'SoLuong'   => self::parseNumber(self::text($hh, 'SLuong')),
                'DonGia'    => self::parseNumber(self::text($hh, 'DGia')),
                'ThanhTien' => self::parseNumber(
                    self::text($hh, 'ThTien') ?: self::text($hh, 'TgTKhau')
                ),
                'VAT'       => self::parseNumber(self::text($hh, 'TSuat')),
                'TienVat'   => self::parseNumber(
                    self::text($hh, 'TienThue') ?: self::text($hh, 'TTSuat')
                ),
                'GhiChu'    => self::text($hh, 'GhiChu'),
            ];
        }

        return ['header' => $header, 'items' => $items];
    }

    // -----------------------------------------------------------------------
    // Helper: lấy text của node con
    // -----------------------------------------------------------------------

    private static function text(SimpleXMLElement $node, string $tag, string $default = ''): string
    {
        $child = $node->{$tag} ?? null;
        if ($child === null) {
            // Thử xpath để tìm bất kể cấp nào
            $found = $node->xpath(".//$tag");
            if (!empty($found)) {
                $child = $found[0];
            }
        }
        if ($child !== null) {
            $val = trim((string) $child);
            if ($val !== '') {
                return $val;
            }
        }
        return $default;
    }

    // -----------------------------------------------------------------------
    // Helper: parse số thực
    // -----------------------------------------------------------------------

    public static function parseNumber(?string $value)
    {
        if ($value === null || $value === '') {
            return null;
        }
        $clean = str_replace(',', '.', $value);
        return is_numeric($clean) ? (float) $clean : null;
    }

    // -----------------------------------------------------------------------
    // Helper: parse số nguyên
    // -----------------------------------------------------------------------

    private static function parseInt(?string $value): ?int
    {
        if ($value === null || $value === '') {
            return null;
        }
        return is_numeric($value) ? (int) $value : null;
    }

    // -----------------------------------------------------------------------
    // Helper: xoá namespace khỏi tất cả node trong DOM
    // -----------------------------------------------------------------------

    private static function stripNamespaces(DOMElement $element): void
    {
        // Đổi tên tag bỏ namespace prefix
        if (strpos($element->tagName, ':') !== false) {
            [, $localName] = explode(':', $element->tagName, 2);
            $newElement = $element->ownerDocument->createElement($localName);
            // Copy attributes (bỏ xmlns:*)
            foreach ($element->attributes as $attr) {
                if (strpos($attr->name, 'xmlns') !== 0) {
                    $newElement->setAttribute($attr->name, $attr->value);
                }
            }
            // Move children
            while ($element->firstChild) {
                $newElement->appendChild($element->firstChild);
            }
            $element->parentNode->replaceChild($newElement, $element);
            $element = $newElement;
        } else {
            // Remove xmlns attributes
            $attrsToRemove = [];
            foreach ($element->attributes as $attr) {
                if (strpos($attr->name, 'xmlns') === 0) {
                    $attrsToRemove[] = $attr->name;
                }
            }
            foreach ($attrsToRemove as $attrName) {
                $element->removeAttribute($attrName);
            }
        }
        // Recurse vào children
        $children = iterator_to_array($element->childNodes);
        foreach ($children as $child) {
            if ($child instanceof DOMElement) {
                self::stripNamespaces($child);
            }
        }
    }
}
